liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -228,32 +228,50 @@ public void stopServer() {
-
-        // check default directory for war.xml (apps or dropins)
-        boolean looseApp = false;
-        File appsDestDir = new File(serverDirectory, "apps");
-        File dropinsDestDir = new File(serverDirectory, "dropins");
-        if (warXmlExists(appsDestDir) || warXmlExists(dropinsDestDir)){
-            looseApp = true;
-        }
-        if (!looseApp){
-            log.info("Installing the application in loose application mode.");
-            runMojo("install-apps", serverName, null);
+        // collect artifacts absolute paths in order to build classpath
+        Set<Artifact> artifacts = project.getArtifacts();
+        List<String> artifactPaths = new ArrayList<String>();
+        for (Artifact artifact : artifacts) {
+            artifactPaths.add(artifact.getFile().getAbsolutePath());
         }
         
-        util = new DevMojoUtil(jvmOptions, serverDirectory);
-
+        // create an executor for tests with an additional queue of size 1, so any further changes detected mid-test will be in the following run
+        final ThreadPoolExecutor executor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS,
+                new ArrayBlockingQueue<Runnable>(1, true));
+        
+        log.info("Running maven-compiler-plugin:compile");
+        runMojo("org.apache.maven.plugins:maven-compiler-plugin", "compile", null, null);
+        log.info("Running maven-compiler-plugin:resources");
+        runMojo("org.apache.maven.plugins:maven-resources-plugin", "resources", null, null);
+        log.info("Running maven-compiler-plugin:testCompile");
+        runMojo("org.apache.maven.plugins:maven-compiler-plugin", "testCompile", null, null);
+        log.info("Running maven-compiler-plugin:testResources");
+        runMojo("org.apache.maven.plugins:maven-resources-plugin", "testResources", null, null);
         sourceDirectory = new File(sourceDirectoryString.trim());
         testSourceDirectory = new File(testSourceDirectoryString.trim());
+        
+        ArrayList<File> javaFiles = new ArrayList<File>();
+        listFiles(sourceDirectory, javaFiles, ".java");
+        
+        ArrayList<File> javaTestFiles = new ArrayList<File>();
+        listFiles(testSourceDirectory, javaTestFiles, ".java");
 
         log.debug("Source directory: " + sourceDirectory);
         log.debug("Output directory: " + outputDirectory);
         log.debug("Test Source directory: " + testSourceDirectory);
         log.debug("Test Output directory: " + testOutputDirectory);
 
-        // create an executor for tests with an additional queue of size 1, so any further changes detected mid-test will be in the following run
-        final ThreadPoolExecutor executor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS,
-                new ArrayBlockingQueue<Runnable>(1, true));
+        recompileJavaSource(javaFiles, artifactPaths, executor, true);
+        recompileJavaTest(javaTestFiles, artifactPaths, executor, true);
```

Comment:
Calling recompileJavaSource and recompileJavaTest would each run the tests.  Line 315 runs the tests on startup as well.  Does this result in the tests running 3 times on startup?